### PR TITLE
Sync: Better Post type syncing

### PR DESCRIPTION
**Pull request by enejb**
- Add a whitelist to what we sync.
- Add a way to convert the synced data back to what we expect
- Add tests to show that things work as expected.

Fixes p4IuY0-TS-p2
#### Changes proposed in this Pull Request:
* Only sync sanitized post type data.

#### Testing instructions:
* Do the tests pass? 
* Do we get expected values on the .com side? 
* Do subscription emails, publicize and other things work as expected?

### CODE REVIEW

**Comment by enejb** [comment]
@lezama can you take a look at this?

File: sync/class.jetpack-sync-defaults.php [hunk 1]
```diff
@@ -195,6 +195,39 @@ public static function get_constants_whitelist() {
...
+		'hierarchical'        => false,
+		'exclude_from_search' => true,
+		'publicly_queryable'  => null,
+		'show_ui'             => false,
+		'show_in_menu'        => null,
+		'show_in_nav_menus'   => null,
+		'show_in_admin_bar'   => false,
+		'menu_position'       => null,
+		'menu_icon'           => null,
+		'supports'			  => array(),
```
Comment (lezama) [inline sync/class.jetpack-sync-defaults.php]: indentation 💥

File: sync/class.jetpack-sync-defaults.php [hunk 2]
```diff
@@ -195,6 +195,39 @@ public static function get_constants_whitelist() {
...
+		'capability_type'     => 'post',
+		'capabilities'        => array(),
+		'cap'				  => array(),
```
Comment (lezama) [inline sync/class.jetpack-sync-defaults.php]: indentation 💥

File: sync/class.jetpack-sync-functions.php [hunk 3]
```diff
@@ -66,7 +66,38 @@ public static function sanitize_taxonomy( $taxonomy ) {
...
+			} else {
+				error_log( 'Jetpack: Encountered a recusive post_type:' . $post_type_name );
+			}
+		}
+		return $post_types_without_callbacks;
+	}
+
+	public static function sanitize_post_type( $post_type ) {
+		// Lets clone the post type object instead of modifing the global one.
+		$cloned_post_type = json_decode( wp_json_encode( $post_type ) );
```
Comment (lezama) [inline sync/class.jetpack-sync-functions.php]: I think we can skip the cloning part as we are assigning everything to a new array.

File: sync/class.jetpack-sync-functions.php [hunk 4]
```diff
@@ -66,7 +66,38 @@ public static function sanitize_taxonomy( $taxonomy ) {
...
 		global $wp_post_types;
 
-		return $wp_post_types;
+		$post_types_without_callbacks = array();
+		foreach ( $wp_post_types as $post_type_name => $post_type ) {
+			$sanitized_post_type = self::sanitize_post_type( $post_type );
+			if ( ! empty( $sanitized_post_type ) ) {
+				$post_types_without_callbacks[ $post_type_name ] = $sanitized_post_type;
+			} else {
+				error_log( 'Jetpack: Encountered a recusive post_type:' . $post_type_name );
```
Comment (lezama) [inline sync/class.jetpack-sync-functions.php]: Maybe we could use `Jetpack::log`?

Comment (enejb) [reply sync/class.jetpack-sync-functions.php]: Lets keep it like this for now.

**Pull request merged**